tree/tree/src/InternalTreeUtils.cxx

```diff
@@ -0,0 +1,175 @@
+         // via chain.AddFile(file.root/myTree) and as well if the tree name is given
+         // in the constructor via TChain(myTree) and a file is added later by chain.AddFile(file.root).
+
+         // Get name of the trees building the chain
+         const auto *chainFiles = static_cast<const TChain *>(frTree)->GetListOfFiles();
+         const auto *realName = chainFiles->First()->GetName();
+         friendNames.emplace_back(std::make_pair(realName, alias));
+         // Get filenames stored in the title member
+         for (const auto *f : *chainFiles) {
+            fileNames.emplace_back(f->GetTitle());
+         }
+      } else {
+         // Get name of the tree
+         const auto realName = GetTreeFullPaths(*frTree)[0];
+         friendNames.emplace_back(std::make_pair(realName, alias));
+
+         // Get filename
+         const auto *f = frTree->GetCurrentFile();
+         if (!f)
+            throw std::runtime_error("Friend trees with no associated file are not supported.");
+         fileNames.emplace_back(f->GetName());
+      }
+   }
+
+   return FriendInfo{std::move(friendNames), std::move(friendFileNames)};
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// \fn std::vector<std::string> GetTreeFullPaths(const TTree &tree)
+/// \ingroup tree
+/// \brief Retrieve the full path(s) to a TTree or the trees in a TChain.
+/// \param[in] tree The tree or chain from which the paths will be retrieved.
+/// \return If the input argument is a TChain, returns a vector of strings with
+///         the name of the tree of each file in the chain. If the input
+///         argument is a TTree, returns a vector with a single element that is
+///         the full path of the tree in the file (e.g. the name of the tree
+///         itself or the path with the directories inside the file). Finally,
+///         the function returns just the name of the tree if it couldn't do any
+///         better.
+std::vector<std::string> GetTreeFullPaths(const TTree &tree)
+{
+   // Case 1: this is a TChain. For each file it contains, GetName returns the name of the tree in that file
+   if (tree.IsA() == TChain::Class()) {
+      const auto &chain = static_cast<const TChain &>(tree);
+      const auto *files = chain.GetListOfFiles();
+      if (!files || files->GetEntries() == 0) {
+         throw std::runtime_error("Input TChain does not contain any file");
+      }
+      std::vector<std::string> treeNames;
+      for (const auto *f : *files)
+         treeNames.emplace_back(f->GetName());
+
+      return treeNames;
+   }
+
+   // Case 2: this is a TTree: we get the full path of it
+   if (const auto *motherDir = tree.GetDirectory()) {
+      // We have 2 subcases (ROOT-9948):
+      // - 1. motherDir is a TFile
+      // - 2. motherDir is a directory
+      // If 1. we just return the name of the tree, if 2. we reconstruct the path
+      // to the file.
+      if (motherDir->InheritsFrom("TFile")) {
```

Comment:
```suggestion
      if (motherDir == motherDIr->GetFile()) {
```

Reply:
can you please add a comment about the magic here? :smile:

Reply:
I think that's because [TDirectory::GetFile](https://github.com/root-project/root/blob/f75a7c464b8b3f49b4ad1ae0e536d139235863d2/core/base/inc/TDirectory.h#L211) returns 0 while [TDirectoryFile::GetFile](https://github.com/root-project/root/blob/f75a7c464b8b3f49b4ad1ae0e536d139235863d2/io/io/inc/TDirectoryFile.h#L91) returns the pointer to the file. But I would propose to use `dynamic_cast<TFile *>` here as well since I find it clearer

Reply:
(I have no idea whether `dynamic_cast<TFile *>(motherDir)` and `motherDir == motherDir->GetFile()` have the same effect)